### PR TITLE
Final project - Back end

### PR DESCRIPTION
**Pull request by PolinaStamenova**
## [Front end PR](https://github.com/PolinaStamenova/ride-fast-front-end/pull/11)

# Final project - Back end

### Database:
**Usage: `PostgreSql`.**
**Created tables**:  `user`, `cars`, `reservations`, `cities`.

### Models:
- All associations are added to the models, so the relations made into the database between the tables, can be used in rails as well. 
- All validations are added to the models, in order to avoid incorrect "inputs values"  from a user into the database.

### Controllers:
- **Users**:  
  - `:create` is used to add a new object of user to the database. 
  - `:destroy` is used to delete a user from the database. 
  - `private method` that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information. 
 
 - **Cars**:  
   - `:index` is used to list all the cars from the database. 
   - `:create` is used to add a new object of car to the database. 
   - `:destroy` is used to delete a car from the database. 
   - `private method` that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information. 

- **Reservations**:  
  - `:index` is used to list all the reservations that a current User has. 
  - `:create` is used to add a new object of reservation to the database. 
  - `:destroy` is used to delete a reservation from the database. 
  - `private method` that ensures that a specific parameter is present and avoid penetration into the project. As someone from outside to get an id of user and get personal information. 

### Routes:
- The routes are created for each method in the controllers, in order the to create endpoints of the Api. 

### API Documentation:
- API Documentation is created, so the Api can be easily used by users. 

### Contributors:
@Ghiftee 
@vikitaotiz
@PolinaStamenova 
@hiromataba 

_Thank you for taking a time to review our PR!_ 😃 🌵

### CODE REVIEW

File: app/controllers/api/v1/auth_controller.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+class Api::V1::AuthController < ApplicationController
+  def signin
+    @user = User.where(username: params[:username]).first
+
+    if @user
+      @user.isLoggedIn = true
```
Comment (MuriloRoque) [inline app/controllers/api/v1/auth_controller.rb]: - [x] You are setting `isLoggedIn` param to `@user` but you are not saving it, so this property is useless in your database. I suggest that you either use `update` here or remove it from the database entirely and just send a `isLoggedIn: true` property in the JSON.

File: app/controllers/api/v1/auth_controller.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      render json: { message: 'User already exists' }, status: :unauthorized
+
+    end
+  end
+
+  def signout
+    @user = User.find(params[:user_id])
+
+    if @user
+      @user.isLoggedIn = false
```
Comment (MuriloRoque) [inline app/controllers/api/v1/auth_controller.rb]: - [x] Same as above.

**Review by MuriloRoque — CHANGES_REQUESTED** [review]
Hi @PolinaStamenova, @vikitaotiz, @Ghiftee, @hiromataba, 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better.

You can also consider:
- Add tests to your back-end code.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Comment by PolinaStamenova** [comment]
Hi @MuriloRoque , 

Thank you for the effort and your suggestions! 😃  🙏

File: app/controllers/api/v1/auth_controller.rb (hunk 1 above)
Comment (PolinaStamenova) [reply app/controllers/api/v1/auth_controller.rb]: Issue fixed

File: app/controllers/api/v1/auth_controller.rb (hunk 2 above)
Comment (PolinaStamenova) [reply app/controllers/api/v1/auth_controller.rb]: Issue fixed

**Review by MuriloRoque — CHANGES_REQUESTED** [review]
Hi @PolinaStamenova, @vikitaotiz, @Ghiftee, @hiromataba, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there is still one which isn't addressed yet.

## Suggested changes

- [ ] Add tests for your back-end.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Comment by PolinaStamenova** [comment]
Hi @MuriloRoque, 

we agree the test are important, but they are not part of the requirements. 
Could you please provide us with information (screenshots) where it says we have to do tests ! And we will gladly create them. 

Thank you a lot in advance! :)

**Review by MuriloRoque — APPROVED** [review]
Hi @PolinaStamenova, @vikitaotiz, @Ghiftee, @hiromataba, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

Good luck! 🎉 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Comment by Ghiftee** [comment]
Thank you for the review and your time@MuriloRoque

**Pull request merged**